test/helpers/value_hook.dart

```diff
@@ -1,42 +1,198 @@
-/// calledHandler.expect(equals(true));
+/// // If the handler was not called after 5s, throw an error
+/// calledHandler.awaitCompletion(Duration(seconds: 5));
 /// ```
 class VHook<T> {
   /// The value of the variable.
-  T _value;
+  dynamic _value;
+
+  /// The error that might be completed with
+  Object _error = _VHookValue.none;
+
+  /// Flag that indicates that the value reached it's final state.
+  final _completer = Completer<T>();
+
+  /// Stream used to subscribe to value changes
+  ///
+  /// Used to allow awaiting values with [awaitValue]
+  final _stream = StreamController<T>.broadcast();
 
   /// Construct a new ValueHook with the given initial value.
-  VHook(this._value);
+  ///
+  /// If not initial value is specified, a placeholder will
+  /// be used. Please check [hasValue] before directly accessing
+  /// [value] in such a case.
+  VHook([this._value = _VHookValue.none]);
 
   /// Set the value of the variable.
-  void set(T val) => _value = val;
+  void set(T val) {
+    // Do not change value if already completed
+    _assertNotCompleted();
+
+    // Update value
+    _value = val;
+
+    // Notify listeners
+    _stream.add(val);
+  }
+
+  /// Complete with the given value.
+  void completeValue(T val) {
+    // Update value
+    set(val);
+
+    // Complete with newly updated value
+    complete();
+  }
+
+  /// Complete with the current value.
+  void complete() {
+    // Can not complete multiple times
+    _assertNotCompleted();
+
+    // Close stream
+    _stream.close();
+
+    // Complete value
+    _completer.complete(_value);
+  }
+
+  /// Complete with an error or exception
+  void completeError(Object error) {
+    // Can not complete multiple times
+    _assertNotCompleted();
+
+    // Save error
+    _error = error;
+
+    // Publish error to all listeners
+    _stream.addError(error);
+
+    // Complete with error
+    _completer.completeError(error);
+  }
+
+  /// Update value with the return of [updater] which takes
+  /// the current value as its only argument.
+  ///
+  /// The new value is returned.
+  ///
+  /// [updater] will not be called if no value has been set, use
+  /// [orElse] to provide a fallback in such a case. An error will be
+  /// thrown if neither [updater] nor [orElse] can be used.
+  ///
+  /// ```dart
+  /// // Construct VHook with initial value
+  /// VHook<int> myHook = VHook<int>(someInteger);
+  ///
+  /// // Raise the current value (any value) to the power of two (x^2)
+  /// int newVal = myHook.update((int val) => val*val);
+  /// ```
+  T update(T Function(T) updater, {dynamic orElse = _VHookValue.none}) {
+    // Do not change value if already completed
+    _assertNotCompleted();
+
+    if (_value != _VHookValue.none) {
+      // Some value already set -> type guarantee -> can use updater
+      set(updater(_value));
+    } else if (orElse != _VHookValue.none) {
+      // No value set + fallback provided -> use fallback
+      set(orElse);
```

Comment:
because `orElse` is dynamic u could call `set` with a value not of type `T` which would assign `_value` with a type unlike `T`. I guess it would lead to some runtime type error at some point. As already pointed out in the first review comment I'd like to migrate from `dynamic` to something type safe

Reply:
This would throw a runtime error when calling `set(orElse)` as `set` takes one argument of type `T`. We have to keep `orElse` dynamic as we use the default of `_VHookValue.none`. This dynamic is the price we pay for making `null` a first class citizen.